File: scripts/ops/ingest-docs.js
#!/usr/bin/env node
/**
 * Ingest documentation into the agent-service knowledge base.
 *
 * Usage:
 *   node scripts/ops/ingest-docs.js [dir1 dir2 ...]
 *
 * Defaults to documentation/user and documentation/developer if no dirs are
 * passed. Reads JWT_SECRET from .env, mints a short-lived HS256 token, and
 * POSTs each file's content to {AGENT_URL}/agent/ingest.
 */
const fs = require("node:fs");
const path = require("node:path");
const crypto = require("node:crypto");

const ROOT = path.resolve(__dirname, "../..");
const ENV_PATH = path.join(ROOT, ".env");
// AGENT_SERVICE_URL points at the agent-service base. From the host (default),
// Caddy fronts it at /agent. Inside the docker network, set
// AGENT_SERVICE_URL=http://agent-service:3000.
const AGENT_URL = process.env.AGENT_SERVICE_URL || "http://localhost/agent";
const READY_TIMEOUT_MS = Number(process.env.READY_TIMEOUT_MS || 60000);

function loadEnv() {
  const text = fs.readFileSync(ENV_PATH, "utf8");
  const out = {};
  for (const line of text.split("\n")) {
    const m = line.match(/^([A-Z_][A-Z0-9_]*)=(.*)$/);
    if (m) out[m[1]] = m[2].replace(/^"|"$/g, "");
  }
  return out;
}

function b64url(buf) {
  return Buffer.from(buf)
    .toString("base64")
    .replace(/=/g, "")
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}

function signJwt(payload, secret) {
  const header = { alg: "HS256", typ: "JWT" };
  const body =
    b64url(JSON.stringify(header)) + "." + b64url(JSON.stringify(payload));
  const sig = crypto.createHmac("sha256", secret).update(body).digest();
  return body + "." + b64url(sig);
}

function walk(dir) {
  const out = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) out.push(...walk(full));
    else if (/\.(qd|md|markdown)$/i.test(entry.name)) out.push(full);
  }
  return out;
}

function stripQdDirectives(text) {
  // Drop top-level Quarkdown directives (.docname {..}, .include {..}) — they
  // would otherwise show up as opaque chunks. Section headings carry the same
  // info via the chunker's section_path.
  return text.replace(/^\.[a-zA-Z_][\w]*\s*\{[^}]*\}\s*\n?/gm, "");
}

async function waitForReady() {
  const deadline = Date.now() + READY_TIMEOUT_MS;
  let lastErr = "";
  while (Date.now() < deadline) {
    try {
      const r = await fetch(`${AGENT_URL}/health`);
      if (r.ok) return;
      lastErr = `HTTP ${r.status}`;
    } catch (e) {
      lastErr = e.message;
    }
    await new Promise((r) => setTimeout(r, 1500));
  }
  throw new Error(
    `agent-service did not become ready at ${AGENT_URL} within ${READY_TIMEOUT_MS}ms (last: ${lastErr})`,
  );
}

async function ingest(file, token) {
  const content = stripQdDirectives(fs.readFileSync(file, "utf8"));
  const source = path.relative(ROOT, file);
  const res = await fetch(`${AGENT_URL}/ingest`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      cookie: `authentication_token=${token}`,
    },
    body: JSON.stringify({
      source,
      content,
      metadata: { type: "user_doc", path: source },
      permissions: [], // public — visible to all callers
    }),
  });
  const body = await res.text();
  return { source, status: res.status, body };
}

async function main() {
  const env = loadEnv();
  const secret = env.JWT_SECRET;
  if (!secret) {
    console.error("JWT_SECRET not found in .env — run `just stack env` first.");
    process.exit(1);
  }

  const dirs =
    process.argv.slice(2).length > 0
      ? process.argv.slice(2)
      : [
          path.join(ROOT, "documentation/user"),
          path.join(ROOT, "documentation/developer"),
        ];

  const files = dirs.flatMap((d) => (fs.existsSync(d) ? walk(d) : []));
  if (files.length === 0) {
    console.error("No .qd/.md files found in:", dirs.join(", "));
    process.exit(1);
  }

  const token = signJwt(
    {
      accountName: "docs-ingester",
      accountId: "docs-ingester",
      accountMemberships: [],
      iat: Math.floor(Date.now() / 1000),
      exp: Math.floor(Date.now() / 1000) + 600,
    },
    secret,
  );

  console.log(`Waiting for agent-service at ${AGENT_URL} ...`);
  await waitForReady();
  console.log(`Ingesting ${files.length} file(s) into ${AGENT_URL} ...`);
  let ok = 0;
  let skipped = 0;
  let failed = 0;
  for (const f of files) {
    try {
      const { source, status, body } = await ingest(f, token);
      if (status >= 200 && status < 300) {
        const j = JSON.parse(body);
        if (j.skipped) skipped++;
        else ok++;
        console.log(
          `  ${j.skipped ? "↺" : "✓"} ${source}  chunks=${j.chunks ?? "-"}${j.skipped ? " (already ingested)" : ""}`,
        );
      } else {
        failed++;
        console.log(`  ✗ ${source}  HTTP ${status}  ${body.slice(0, 200)}`);
      }
    } catch (e) {
      failed++;
      console.log(`  ✗ ${f}  ${e.message}`);
    }
  }
  console.log(`\nDone. ingested=${ok} unchanged=${skipped} failed=${failed}`);
  if (failed > 0) process.exit(1);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/ops/ingest-docs.py
#!/usr/bin/env python3
"""Ingest documentation into the agent knowledge base: mints a short-lived
HS256 token (see backend/agent/CLAUDE.md) and POSTs each file to {AGENT_URL}/agent/ingest."""

import base64
import hashlib
import hmac
import json
import os
import re
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
ENV_PATH = ROOT / ".env"
AGENT_URL = os.environ.get("AGENT_URL", "http://localhost/agent")
READY_TIMEOUT_S = float(os.environ.get("READY_TIMEOUT_MS", 60000)) / 1000
HEALTH_TIMEOUT_S = 5
INGEST_TIMEOUT_S = 120


def load_env():
    out = {}
    for line in ENV_PATH.read_text(encoding="utf-8").split("\n"):
        match = re.match(r"^([A-Z_][A-Z0-9_]*)=(.*)$", line)
        if match:
            out[match.group(1)] = re.sub(r'^"|"$', "", match.group(2))
    return out


def b64url(raw):
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode()


def sign_jwt(payload, secret):
    header = {"alg": "HS256", "typ": "JWT"}
    dump = lambda obj: json.dumps(obj, separators=(",", ":")).encode()
    body = b64url(dump(header)) + "." + b64url(dump(payload))
    signature = hmac.new(secret.encode(), body.encode(), hashlib.sha256).digest()
    return body + "." + b64url(signature)


def walk(directory):
    out = []
    for entry in sorted(os.listdir(directory)):
        full = os.path.join(directory, entry)
        if os.path.isdir(full):
            out.extend(walk(full))
        elif re.search(r"\.(qd|md|markdown)$", entry, re.IGNORECASE):
            out.append(full)
    return out


def strip_qd_directives(text):
    return re.sub(r"^\.[a-zA-Z_]\w*\s*\{[^}]*\}\s*\n?", "", text, flags=re.MULTILINE)


def wait_for_ready():
    deadline = time.monotonic() + READY_TIMEOUT_S
    last_error = ""
    while time.monotonic() < deadline:
        try:
            with urllib.request.urlopen(f"{AGENT_URL}/health", timeout=HEALTH_TIMEOUT_S) as response:
                if 200 <= response.status < 300:
                    return
                last_error = f"HTTP {response.status}"
        except urllib.error.HTTPError as error:
            last_error = f"HTTP {error.code}"
        except Exception as error:
            last_error = str(error)
        time.sleep(1.5)
    raise RuntimeError(
        f"agent did not become ready at {AGENT_URL} within "
        f"{int(READY_TIMEOUT_S * 1000)}ms (last: {last_error})"
    )


def ingest(file, token):
    content = strip_qd_directives(Path(file).read_text(encoding="utf-8"))
    source = os.path.relpath(file, ROOT)
    payload = json.dumps(
        {
            "source": source,
            "content": content,
            "metadata": {"type": "user_doc", "path": source},
            "permissions": [],
        },
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")
    request = urllib.request.Request(
        f"{AGENT_URL}/ingest",
        data=payload,
        method="POST",
        headers={
            "content-type": "application/json",
            "cookie": f"authentication_token={token}",
        },
    )
    try:
        with urllib.request.urlopen(request, timeout=INGEST_TIMEOUT_S) as response:
            return source, response.status, response.read().decode("utf-8", "replace")
    except urllib.error.HTTPError as error:
        return source, error.code, error.read().decode("utf-8", "replace")


def main():
    secret = load_env().get("EVAL_JWT_SECRET")
    if not secret:
        print("EVAL_JWT_SECRET not found in .env — run `just stack env` first.", file=sys.stderr)
        sys.exit(1)

    dirs = sys.argv[1:] or [str(ROOT / "documentation/user"), str(ROOT / "documentation/developer")]
    files = [f for d in dirs if os.path.exists(d) for f in walk(d)]
    if not files:
        print("No .qd/.md files found in:", ", ".join(dirs), file=sys.stderr)
        sys.exit(1)

    now = int(time.time())
    token = sign_jwt(
        {"sub": "docs-ingester", "roles": [], "domains": [], "iat": now, "exp": now + 600},
        secret,
    )

    print(f"Waiting for agent at {AGENT_URL} ...")
    wait_for_ready()
    print(f"Ingesting {len(files)} file(s) into {AGENT_URL} ...")

    ok = skipped = failed = 0
    for file in files:
        try:
            source, status, body = ingest(file, token)
            if 200 <= status < 300:
                data = json.loads(body)
                chunks = data.get("chunks")
                if data.get("skipped"):
                    skipped += 1
                else:
                    ok += 1
                mark = "↺" if data.get("skipped") else "✓"
                note = " (already ingested)" if data.get("skipped") else ""
                print(f"  {mark} {source}  chunks={'-' if chunks is None else chunks}{note}")
            else:
                failed += 1
                print(f"  ✗ {source}  HTTP {status}  {body[:200]}")
        except Exception as error:
            failed += 1
            print(f"  ✗ {file}  {error}")

    print(f"\nDone. ingested={ok} unchanged={skipped} failed={failed}")
    if failed:
        sys.exit(1)


try:
    main()
except SystemExit:
    raise
except Exception as error:
    print(error, file=sys.stderr)
    sys.exit(1)
